core/src/main/scala/org/apache/spark/Partition.scala

```diff
@@ -26,6 +26,11 @@ trait Partition extends Serializable {
    */
   def index: Int
 
+  override def equals(other: Any): Boolean = other match {
+    case that: Partition => super.equals(that) && that.index == index
```

Comment:
Actually is there any subclass that relies on this default implementation? If so, I think it also needs to check its own class vs the class of the argument. If not, we could remove this.